cgp/genome.py

```diff
@@ -267,6 +267,179 @@ def randomize(self, rng: np.random.RandomState) -> None:
+        """
+        addable_nodes = self._get_addable_nodes(node_dependencies)
+        used_node_indices.append(current_idx)
+        addable_nodes = addable_nodes.difference(used_node_indices)
+
+        return addable_nodes, used_node_indices
+
+    def _update_gene_values(self, dna: List[int], used_node_indices: List[int]) -> None:
+        """ Update gene values to correspond to the new position of the corresponding node"""
+        for gene_idx, gene_value in enumerate(dna):
+            region_idx = self._get_region_idx(gene_idx)
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                if gene_value >= self._n_inputs:
+                    gene_value = self._n_inputs + used_node_indices.index(gene_value)
+            dna[gene_idx] = gene_value
+
+    def _replace_invalid_gene_values(
+        self, dna_for_reorder: List[int], rng: np.random.RandomState
+    ) -> None:
+        """ Replace gene values of unused arities
+        by (pseudo-) random picks of permissible values for that node.
+        Works only if self.n_rows==1
+        """
+
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = self._get_region_idx(gene_idx)
+            if self._is_hidden_input_gene(gene_idx, region_idx) and gene_value > region_idx:
+                # replace value that is too large by valid value
+                permissible_values = self.determine_permissible_values_per_gene(gene_idx)
+                gene_value = rng.choice(permissible_values)
+                dna_for_reorder[gene_idx] = gene_value
+
+    def _get_addable_nodes(self, node_dependencies: Dict[int, set]) -> Set[int]:
+
+        return set(
+            idx for idx, dependencies in node_dependencies.items() if len(dependencies) == 0
+        )
+
+    def _get_region_idx(self, gene_idx: int) -> int:
+
+        return gene_idx // self._length_per_region
+
+    def _determine_node_dependencies(self) -> Dict[int, set]:
+        """ Determines for every node a set of dependencies.
+
+            The set of dependencies for a node is given by the indices of the nodes,
+            on which its inputs depend.
+            The arity of the operator gene is taken into account, only nodes which
+            are used by the operator are seen as dependency.
+            Implicit dependencies are not added, since they are covered via the
+            set of explicit dependencies of the respective node.
+
+
+        Returns
+        ----
+        dependencies: Dict
+            Dictionary containing for every node a set of internal nodes dependencies
+
+        """
+        dependencies = {}
+        for region_idx in range(self._n_regions):
+
+            if self._is_input_region(region_idx) or self._is_output_region(region_idx):
+                continue
+
+            elif self._is_hidden_region(region_idx):
```

Comment:
couldn't this be simplified by using the `_iter_hidden_regions` iterator?

Reply:
Yes, makes it much simpler to understand!